modules/core/arrow-core-data/src/main/kotlin/arrow/typeclasses/Traverse.kt

```diff
@@ -5,11 +5,519 @@ import arrow.core.Id
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val list = listOf("1", "2", "3").k().traverse(Either.applicative(), ::parseIntEither)
+ *   val failFastList = listOf("1", "abc", "3", "4s").k().traverse(Either.applicative(), ::parseIntEither)
+ *   //sampleEnd
+ *   println("list= $list")
+ *   println("failFastList= $failFastList")
+ * }
+ * ```
+ *
+ * ```kotlin:ank:playground
+ * import arrow.core.Either
+ * import arrow.core.Nel
+ * import arrow.core.ValidatedNel
+ * import arrow.core.extensions.nonemptylist.semigroup.semigroup
+ * import arrow.core.extensions.validated.applicative.applicative
+ * import arrow.core.fix
+ * import arrow.core.invalidNel
+ * import arrow.core.k
+ * import arrow.core.validNel
+ *
+ * fun parseIntEither(s: String): Either<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) Either.Right(s.toInt())
+ *   else Either.Left(NumberFormatException("$s is not a valid integer."))
+ *
+ * fun parseIntValidated(s: String): ValidatedNel<NumberFormatException, Int> =
+ *   if (s.matches(Regex("-?[0-9]+"))) s.toInt().validNel()
+ *   else NumberFormatException("$s is not a valid integer.").invalidNel()
+ *
+ * fun main() {
+ *   //sampleStart
+ *   val validatedList = listOf("1", "22w", "3", "33s").k()
+ *     .traverse(ValidatedNel.applicative(Nel.semigroup<NumberFormatException>()), ::parseIntValidated).fix()
+ *   //sampleEnd
+ *   println(validatedList)
+ * }
+ * ```
+ *
+ * Notice that in the [Either] case, should any string fail to parse the entire [traverse] is considered a failure. Moreover, once it hits its first bad parse, it will not attempt to parse any others down the line (similar behavior would be found with using `Option`). Contrast this with `Validated` where even if one bad parse is hit, it will continue trying to parse the others, accumulating any and all errors as it goes. The behavior of [traverse] is closely tied with the [Applicative] behavior of the data type, where computations are run in isolation.
+ *
+ * Going back to our `IO` example from the beginning with concurrency in mind, we can get an [Applicative] instance for `IO`, that will not short-circuit, by using `parApplicative`.
+ *
+ * It is worth mentioning that `parApplicative` does not implement `lazyAp` in contrast to `IO.applicative()`. `parApplicative` creates all IO's upfront and executes them in parallel. More importantly, `parApplicative` breaks monad-applicative-consistency laws by design. Consequently, `IO.applicative()` will only run an effect, when the previous one is successful.
```

Comment:
```suggestion
 * It is worth mentioning that `parApplicative` does not implement `lazyAp` in contrast to `IO.applicative()`, which means `parApplicative` creates all IO's upfront and executes them in parallel. If one fails, the resulting `IO` fails, but the effects may still have already run. As a consequence `parApplicative` breaks monad-applicative-consistency laws which is by design. On the flipside, `IO.applicative()` will only run an effect, when the previous one is successful.
```
I think this is clearer. Having implemented `lazyAp` I know what it does and what implications it holds, but a first time reader won't. This shows the significance of it in contrast to the normal applicative.

Reply:
I phrased it a bit differently but it catches your point